src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs

```diff
@@ -888,51 +888,48 @@ private void DetermineAndAddExecutablePathCommand(IList<LaunchCommand> commands,
+            }
+        }
+
         private void AddGetTargetArchitectureCommand(IList<LaunchCommand> commands)
         {
-            if (_launchOptions.TargetArchitecture == TargetArchitecture.Unknown)
+            // User may specify the wrong architecture, e.g. ARM instead of ARM64, so use the target's real architecture if available:
+            // 1. if the command factory can discover the target architecture then use that
+            // 2. else if the user specified an architecture then use that
+            // 3. otherwise default to x64
+            TargetArchitecture arch = DefaultArch();
+
+            Task successHandler(string resultsStr)
             {
-                Action<string> failureHandler = (string miError) =>
-                {
-                    string message = ResourceStrings.Error_FailedToGetTargetArchitecture;
-                    throw new LaunchErrorException(message);
-                };
+                var archFromTarget = MICommandFactory.ParseTargetArchitectureResult(resultsStr);
 
-                Func<string, Task> successHandler = (string resultsStr) =>
+                if (archFromTarget != TargetArchitecture.Unknown)
                 {
-                    TargetArchitecture arch = MICommandFactory.ParseTargetArchitectureResult(resultsStr);
-
-                    if (LaunchOptions.TargetArchitecture != TargetArchitecture.Unknown)
-                    {
-                        arch = LaunchOptions.TargetArchitecture;
-                    }
-                    else if (arch == TargetArchitecture.Unknown)
-                    {
-                        // Use X64 as default if the arch couldn't be detected and wasn't specified
-                        // in the launch options
-                        WriteOutput(ResourceStrings.Warning_UsingDefaultArchitecture);
-                        arch = TargetArchitecture.X64;
-                    }
+                    arch = archFromTarget;
+                }
 
-                    SetTargetArch(arch);
+                SetTargetArch(arch);
 
-                    return Task.FromResult(0);
-                };
+                return Task.FromResult(0);
+            }
 
-                string cmd = MICommandFactory.GetTargetArchitectureCommand();
+            string cmd = MICommandFactory.GetTargetArchitectureCommand();
 
-                if (cmd != null)
-                {
-                    commands.Add(new LaunchCommand(cmd, ignoreFailures: false, successHandler: successHandler, failureHandler: failureHandler));
-                }
-                else
-                {
-                    SetTargetArch(MICommandFactory.ParseTargetArchitectureResult(""));
-                }
-            }
-            else
+            if (cmd != null)
             {
-                SetTargetArch(_launchOptions.TargetArchitecture);
+                commands.Add(new LaunchCommand(cmd, ignoreFailures: false, successHandler: successHandler));
```

Comment:
Should we ignore failures if the user specified a value? (in case there are GDBs out there that don't like to say)